### PR TITLE
Hide overlapping x-axis tick labels

### PR DESCRIPTION
**Pull request by krkshitij**
## Previous Behavior

The default number of x-axis ticks causes tick labels to overlap in smaller viewports.

## New Behavior

When the hideTickOverlap prop is set to true, continuous axes (such as number and date) automatically adjust their tick count based on the container width, while categorical/string axes hide any tick labels that might overlap.

### CODE REVIEW

File: packages/react-components/react-charts-preview/library/src/utilities/utilities.ts [hunk 1]
```diff
@@ -300,31 +329,63 @@ export function createStringXAxis(
...
+
   const xAxis = d3AxisBottom(xAxisScale)
     .tickSize(xAxistickSize)
     .tickPadding(tickPadding)
-    .ticks(xAxisCount)
-    .tickFormat((x: string, index: number) => {
-      return convertToLocaleString(dataset[index], culture) as string;
-    });
+    .tickValues(tickValues)
+    .tickFormat(tickFormat);
```
Comment (AtishayMsft) [inline packages/react-components/react-charts-preview/library/src/utilities/utilities.ts:383]: ToDo: add unit tests for this logic

File: packages/react-components/react-charts-preview/library/src/utilities/utilities.ts [hunk 2]
```diff
@@ -244,42 +257,58 @@ export function createDateXAxis(
...
     .domain([domainNRangeValues.dStartValue, domainNRangeValues.dEndValue])
     .range([domainNRangeValues.rStartValue, domainNRangeValues.rEndValue]);
-  const xAxis = d3AxisBottom(xAxisScale).tickSize(xAxistickSize).tickPadding(tickPadding).ticks(xAxisCount);
-  if (customDateTimeFormatter) {
-    xAxis.tickFormat((domainValue: Date, _index: number) => {
+
+  let tickCount = xAxisCount ?? 6;
+  const tickFormat = (domainValue: Date, _index: number) => {
+    if (customDateTimeFormatter) {
       return customDateTimeFormatter(domainValue);
```
Comment (AtishayMsft) [inline packages/react-components/react-charts-preview/library/src/utilities/utilities.ts:276]: we need to ensure user provided customDateTimeFormatter takes priority over other ways to format the ticks

File: packages/react-components/react-charts-preview/library/src/utilities/utilities.ts [hunk 3]
```diff
@@ -244,42 +257,58 @@ export function createDateXAxis(
...
+    if (culture === undefined && tickParams.tickFormat) {
+      if (useUTC) {
+        return d3UtcFormat(tickParams.tickFormat)(domainValue);
+      } else {
+        return d3TimeFormat(tickParams.tickFormat)(domainValue);
+      }
+    }
+    return multiFormat(domainValue, undefined, useUTC);
+  };
+  if (hideTickOverlap && typeof xAxisCount === 'undefined') {
```
Comment (AtishayMsft) [inline packages/react-components/react-charts-preview/library/src/utilities/utilities.ts:294]: ensure that the order of priority defined by the user does not get overridden

Comment (krkshitij) [reply packages/react-components/react-charts-preview/library/src/utilities/utilities.ts:294]: checked it. any user defined customizations take priority over those requiring D3 utils.

File: packages/react-components/react-charts-preview/library/src/utilities/utilities.ts (hunk 1 above)
Comment (krkshitij) [reply packages/react-components/react-charts-preview/library/src/utilities/utilities.ts:383]: Yes, will add the unit tests once the logic is finalized. Still have to cover scenarios like rotation, wrapping and truncation of tick labels. Also noticed that the calculateLongestLabelWidth util is not very accurate.

**Comment by AtishayMsft** [comment]
Could you resolve the conflict of change in prop name in upstream branch

**Comment by AtishayMsft** [comment]
There is another conflict being reported. Could you resolve it.

**Review by AtishayMsft — APPROVED** [review]

**Pull request merged**